### PR TITLE
FACT-2753-parallel-semaphore

### PR DESCRIPTION
**Pull request by IrimieBogdan**
Resolve facts by default in parallel (concurrently with MRI ruby). Sequential resolution of facts can be enables via the --sequential cli flag or via facter.conf (via de global.sequantial)

The semaphore is extracted in the base class compared to #2051

### CODE REVIEW

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

File: lib/facter/framework/core/fact/internal/internal_fact_manager.rb [hunk 1]
```diff
@@ -47,5 +55,40 @@ def resolve(searched_facts)
...
+    def start_threads(searched_facts)
+      threads = []
+      # only resolve a fact once, even if multiple search facts depend on that fact
+      searched_facts
+        .uniq { |searched_fact| searched_fact.fact_class.name }
+        .each do |searched_fact|
+        threads << Thread.new do
+          resolve_fact(searched_fact)
+        end
+      end
```
Comment (gimmyxd) [inline lib/facter/framework/core/fact/internal/internal_fact_manager.rb:68]: you can replace `each` with `map`

```ruby
searched_facts
  .uniq { |searched_fact| searched_fact.fact_class.name }
  .map do |searched_fact|
    Thread.new do
      resolve_fact(searched_fact)
    end
  end
```

File: lib/facter/framework/core/fact/internal/internal_fact_manager.rb [hunk 2]
```diff
@@ -47,5 +55,40 @@ def resolve(searched_facts)
...
+
+      threads
+    end
+
+    def join_threads(threads, searched_facts)
+      resolved_facts = []
+
+      threads.each do |thread|
+        thread.join
```
Comment (gimmyxd) [inline lib/facter/framework/core/fact/internal/internal_fact_manager.rb:77]: i think Thread#join is not needed as Thread#value(https://ruby-doc.org/core-2.5.1/Thread.html#method-i-value) is using join.

File: lib/facter/framework/core/fact/internal/internal_fact_manager.rb [hunk 3]
```diff
@@ -47,5 +55,40 @@ def resolve(searched_facts)
...
+        resolved_facts << thread.value unless thread.value.nil?
+      end
+
+      resolved_facts.flatten!
+
+      FactAugmenter.augment_resolved_facts(searched_facts, resolved_facts)
```
Comment (gimmyxd) [inline lib/facter/framework/core/fact/internal/internal_fact_manager.rb:83]: Might be able to write this shorter as
```ruby
resolved_facts = threads.map(&:value)
resolved_facts.compact!

FactAugmenter.augment_resolved_facts(searched_facts, resolved_facts)
```

File: lib/facter/framework/core/fact/internal/internal_fact_manager.rb (hunk 2 above)
Comment (IrimieBogdan) [reply lib/facter/framework/core/fact/internal/internal_fact_manager.rb]: Removed join

File: lib/facter/framework/core/fact/internal/internal_fact_manager.rb (hunk 1 above)
Comment (IrimieBogdan) [reply lib/facter/framework/core/fact/internal/internal_fact_manager.rb:67]: Replaced each with map 👍

File: lib/facter/framework/core/fact/internal/internal_fact_manager.rb (hunk 3 above)
Comment (IrimieBogdan) [reply lib/facter/framework/core/fact/internal/internal_fact_manager.rb]: Great suggestion, refactored.

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Review by gimmyxd — APPROVED** [review]

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Pull request merged**